### DISCUSSION TITLE
HL7 to many JSON

### DISCUSSION THREAD

**Discussion by igromano** [nextgenhealthcare/connect#6000] [Q&A] [answered]
Hello everyone! I need to create a channel that will receive HL7 v2.6 messages and respond in JSON format. So far, so good. The problem I'm facing is that some of the incoming messages have multiple OBR segments, and I need to create a new JSON for each of them. I iterate through the segments, but the output is a JSON with the data from the last OBR. The question is, how can I generate either a JSON array and process it from another channel, or a new JSON for each OBR and process each of them from another channel?

Thanks

**Comment by pacmano1** [comment]
Post your code attempt, a sample message and your expected output.

Also, your use case is not clear.  You wrote "need to respond", that is a specific technical thing, what do you mean by that?

**Reply by igromano** [reply]
"Thank you for your response! I have a channel that will receive HL7 messages, which can have one or more OBR segments. Here's an example:

**MSH|^~\&|PRE|SMG|RIS|SMG|20231115101117||ORM^O01^ORM_O01|1711558|P|2.3|||AL|||UNICODE UTF-8|| 
PID||547860|25378230^^^1||TEST^TEST TEST ||19000521|F||||||||||||||||||||||||||||| 
PV1|1|H|1^^D-101A|||||||239^DLA Radiología Simple|||||||||4546656-01||||||||||||||||||||||||||||||||V
IN1|||2|SMG
TQ1|1||||||||S||||| 
ORC|NW|G4546656-01||G4546656-01|IP|8000060675992051019|||20231115101103|1043^MEDICO PRUEBA^|||1^DLA Adulto Clínico Quirúrgico||20231115101039|239^DLA Radiología Simple|1^DLA Adulto Clínico Quirúrgico||||||||||||^. prueba^| 
OBR|1|G4546656-01||340301^RADIOGRAFIA DE TORAX FRENTE^||||||||||||||2806|
OBR|2|G4546656-01||340501^RADIOGRAFIA DE ARBOL URINARIO^||||||||||||||2806|**

I need to transform it into JSON messages to send them to a RIS. However, I need to send a JSON for each OBR that exists in the HL7, containing the same medical order data but each one with the procedure specific to each OBR. Here's an example of the JSON output I expect:
{
  "AppointmentDateTime": "date",
  "PatientType": "string",
  "Site": "string",
  "HealthCoverage": "string",
  "RequiresReport": "string",
  "PatientID": {
    "PatientHisID": "string",
    "PatientDocument": "string"
  },
  "Surname": "string",
  "Sex": "string",
  "Name": "string",
  "ProcedureDescription": "string",
  "DOB": "string",
  "Solicitor": "string",
  "mail": "string",
  "HasConsent": "boolean",
  "Mobility": "string",
  "AffiliateNumber": "string",
  "HisId": "string",
  "RequesterRegistrationId": "string",
  "RequesterName": "string",
  "RequesterSurame": "string",
  "RequesterRegistrationType": "string",
  "Flags": "string",
  "DeviceType": "string",
  "ProcedureGroupID": "string",
  "DeviceID": "string",
  "ImageQty": "string"
}

As I mentioned before, I need a JSON for each procedure, and each of these must be sent to a RIS web service using a POST request. What I had originally planned is that each JSON could be sent to another channel responsible for connecting with the RIS.

I'm sharing the little code I've been working on, but I haven't achieved a successful result yet

```var name = msg['PID']['PID.5']['PID.5.2'].toString();
var lastName = msg['PID']['PID.5']['PID.5.1'].toString();
var orderDate = msg['ORC']['ORC.9']['ORC.9.1'].toString();
var insurance =  msg['IN1']['IN1.4']['IN1.4.1'].toString();
var sex = msg['PID']['PID.8']['PID.8.1'].toString();
var dob = msg['PID']['PID.7']['PID.7.1'].toString();
var mail = "";
var affNumber = msg['ORC']['ORC.6']['ORC.6.1'].toString();

var index = 0;

for each(obr in msg['OBR']){
	procedureName = msg['OBR'][index]['OBR.4']['OBR.4.2'].toString();
	hisId = msg['OBR'][index]['OBR.2']['OBR.2.1'].toString() + "-" + msg['OBR'][0]['OBR.1']['OBR.1.1'].toString();
	///MAPPINGS
	tmp['Name'] = name;
	tmp['Surname'] = lastName;
	tmp['AppointmentDateTime'] = orderDate;
	tmp['HealthCoverage'] = insurance;
	tmp['Sex'] = sex;
	tmp['DOB'] = dob;
	tmp['mail'] = mail;
	tmp['AffiliateNumber'] = affNumber;
	tmp['ProcedureDescription'] = procedureName;
	tmp['HisId'] = hisId;
	index++
	//TEST
	router.routeMessage('DEV_0_JSON_to_File', tmp.toString());
}```

**Reply by pacmano1** [reply]
Please fix your post and use code formatting and highlighting, read https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks.

**Reply by pacmano1** [reply, marked as answer]
This is not your complete code, but a template you can work with.   Also, there are a few ways to do this, this is just one.  You are on the right track.

Note that typically you want to send appropriate data types, for example you may want to convert a date to JSON date.  Also, I commented out the `router.routeMessage` statement.

```JS
var this_obr = {}
var obr_list = [] // optional 
for each (obr in msg.OBR) {
    this_obr = {
        PatientID: {
            PatientHisID: msg['PID']['PID.2']['PID.2.1'].toString(),  // this is a hard code
            PatientDocument: "string"
        },
        PlacerField1: obr['OBR.18']['OBR.18.1'].toString()
    }
    // router.routeMessage('SOME CHANNEL THAT ACCEPTS JSON IN AND DOES AN HTTP POST', JSON.stringify(this_obr));  // commented out but you will use this
    obr_list.push(this_obr);
}
$c('obr_list', obr_list) // put to channel map so you can see the JSON array you built
```
When you view message, and look at the value of obr_list you wiil see:

```JSON
[{
    "PatientID": {
        "PatientHisID": "547860",
        "PatientDocument": "string"
    },
    "PlacerField1": "2806"
}, {
    "PatientID": {
        "PatientHisID": "547860",
        "PatientDocument": "string"
    },
    "PlacerField1": "2806"
}]
```

**Reply by pacmano1** [reply]
I had a typo before, it is fixed now.

**Reply by igromano** [reply]
It worked perfectly, thank you very much for the help!
